kernel/src/process/status.rs

```diff
@@ -71,3 +77,97 @@ impl ProcessStatus {
         self.exit_code.store(exit_code, Ordering::Relaxed);
     }
 }
+
+impl ProcessStatus {
+    pub(super) fn stop_status(&self) -> &StopStatus {
+        &self.stop_status
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct StopStatus {
+    /// Indicates whether the process is stopped.
+    is_stopped: AtomicBool,
+
+    /// Indicates whether the process's status has changed and has not yet been waited on.
+    ///
+    /// User programs may use the wait* syscalls to check for changes in
+    /// the process's status. This field will be set to `Some(_)` once the
+    /// process's status changes and will be set to `None` if the process
+    /// has already been waited on.
+    wait_status: SpinLock<Option<StopWaitStatus>>,
+}
+
+impl StopStatus {
+    pub(self) const fn new() -> Self {
+        Self {
+            is_stopped: AtomicBool::new(false),
+            wait_status: SpinLock::new(None),
+        }
+    }
+
+    /// Stops the process by some signal.
+    ///
+    /// The return value indicates whether the stop status has changed.
+    pub(super) fn stop(&self, signum: SigNum) -> bool {
+        // Hold the lock first to avoid race conditions
+        let mut wait_status = self.wait_status.lock();
+
+        if self.is_stopped.load(Ordering::Relaxed) {
+            false
+        } else {
+            self.is_stopped.store(true, Ordering::Relaxed);
+            *wait_status = Some(StopWaitStatus::Stopped(signum));
+            true
+        }
+    }
+
+    /// Resumes the process.
+    ///
+    /// The return value indicates whether the stop status has changed.
+    pub(super) fn resume(&self) -> bool {
+        // Hold the lock first to avoid race conditions
+        let mut wait_status = self.wait_status.lock();
+
+        if self.is_stopped.load(Ordering::Relaxed) {
+            self.is_stopped.store(false, Ordering::Relaxed);
+            *wait_status = Some(StopWaitStatus::Continue);
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Returns whether the process is stopped.
+    pub(super) fn is_stopped(&self) -> bool {
+        self.is_stopped.load(Ordering::Relaxed)
+    }
+
+    /// Waits on the stop status changes.
```

Comment:
I don't think this method actually waits?

```suggestion
    /// Gets and clears the stop status changes for the `wait` syscall.
```

The same for the doc comments of `wait_stopped_or_continued`.